app/src/main/java/org/thoughtcrime/securesms/BlockUnblockDialog.java

```diff
@@ -0,0 +1,127 @@
+
+  private BlockUnblockDialog() { }
+
+  public static void showBlockFor(@NonNull Context context,
+                                  @NonNull Lifecycle lifecycle,
+                                  @NonNull Recipient recipient,
+                                  @NonNull Runnable onBlock)
+  {
+    SimpleTask.run(lifecycle,
+                   () -> buildBlockFor(context, recipient, onBlock, null),
+                   AlertDialog.Builder::show);
+  }
+
+  public static void showBlockAndReportSpamFor(@NonNull Context context,
+                                               @NonNull Lifecycle lifecycle,
+                                               @NonNull Recipient recipient,
+                                               @NonNull Runnable onBlock,
+                                               @NonNull Runnable onBlockAndReportSpam)
+  {
+    SimpleTask.run(lifecycle,
+                   () -> buildBlockFor(context, recipient, onBlock, onBlockAndReportSpam),
+                   AlertDialog.Builder::show);
+  }
+
+  public static void showUnblockFor(@NonNull Context context,
+                                    @NonNull Lifecycle lifecycle,
+                                    @NonNull Recipient recipient,
+                                    @NonNull Runnable onUnblock)
+  {
+    SimpleTask.run(lifecycle,
+                   () -> buildUnblockFor(context, recipient, onUnblock),
+                   AlertDialog.Builder::show);
+  }
+
+  @WorkerThread
+  private static AlertDialog.Builder buildBlockFor(@NonNull Context context,
+                                                   @NonNull Recipient recipient,
+                                                   @NonNull Runnable onBlock,
+                                                   @Nullable Runnable onBlockAndReportSpam)
+  {
+    recipient = recipient.resolve();
+
+    AlertDialog.Builder builder   = new MaterialAlertDialogBuilder(context);
+    Resources           resources = context.getResources();
+
+    if (recipient.isGroup()) {
+      if (DatabaseFactory.getGroupDatabase(context).isActive(recipient.requireGroupId())) {
+        builder.setTitle(resources.getString(R.string.BlockUnblockDialog_block_and_leave_s, recipient.getDisplayName(context)));
+        builder.setMessage(R.string.BlockUnblockDialog_you_will_no_longer_receive_messages_or_updates);
+        builder.setPositiveButton(R.string.BlockUnblockDialog_block_and_leave, ((dialog, which) -> onBlock.run()));
+        builder.setNegativeButton(android.R.string.cancel, null);
+      } else {
+        builder.setTitle(resources.getString(R.string.BlockUnblockDialog_block_s, recipient.getDisplayName(context)));
+        builder.setMessage(R.string.BlockUnblockDialog_group_members_wont_be_able_to_add_you);
+        builder.setPositiveButton(R.string.RecipientPreferenceActivity_block, ((dialog, which) -> onBlock.run()));
+        builder.setNegativeButton(android.R.string.cancel, null);
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 68:78, 107:117.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.